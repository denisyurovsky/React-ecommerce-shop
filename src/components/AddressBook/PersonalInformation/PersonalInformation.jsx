import {
  Box,
  TextField,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
} from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import formatPhoneNumber from '../../../helpers/formatPhoneNumber';
import validateField from '../../../helpers/validateField';
import addressType from '../../../propTypes/addressType';

import styles from './PersonalInformation.module.scss';

const PersonalInformation = ({ handleChange, address, setAddress }) => {
  const { title, name, surname, phone } = address;
  const handlePhoneNumberChange = (e) => {
    const phone = formatPhoneNumber(
      e.target.value.replace(/[^0-9]/g, '').slice(0, 11)
    );

    setAddress({ ...address, phone });
  };

  const [isPhoneNumberValid, setIsPhoneNumberValid] = useState(true);
  const [isTitleValid, setIsTitleValid] = useState(true);
  const [isNameValid, setIsNameValid] = useState(true);
  const [isSurnameValid, setIsSurnameValid] = useState(true);

  const validatePhoneNumber = (e) => {
    const phoneNumberLength = e.target.value.length;

    setIsPhoneNumberValid(phoneNumberLength === 16);
  };

  const validateTitleField = (e) => {
    validateField(e, setIsTitleValid);
  };
  const suspendTitleFieldValidation = () => {
    setIsTitleValid(true);
  };
  const validateNameField = (e) => {
    validateField(e, setIsNameValid);
  };
  const suspendNameFieldValidation = () => {
    setIsNameValid(true);
  };
  const validateSurnameField = (e) => {
    validateField(e, setIsSurnameValid);
  };
  const suspendSurnameFieldValidation = () => {
    setIsSurnameValid(true);
  };
  const suspendPhoneFieldValidation = () => {
    setIsPhoneNumberValid(true);
  };

  return (
    <>
      <Box className={styles.wrapperFullName}>
        <FormControl
          className={styles.title}
          variant="standard"
          onBlur={validateTitleField}
          onFocus={suspendTitleFieldValidation}
          error={!isTitleValid}
        >
          <InputLabel id="title">Title</InputLabel>
          <Select
            data-testid="select-title"
            labelId="title"
            name="title"
            value={title}
            onChange={handleChange}
            label="Title"
          >
            <MenuItem value={'Mr.'}>Mr.</MenuItem>
            <MenuItem value={'Mrs.'}>Mrs.</MenuItem>
          </Select>
        </FormControl>
        <Box className={styles.wrapper}>
          <TextField
            className={styles.name}
            onChange={handleChange}
            id="name"
            value={name}
            label="Name"
            variant="standard"
            onBlur={validateNameField}
            onFocus={suspendNameFieldValidation}
            error={!isNameValid}
            helperText={isNameValid ? ' ' : 'This field should not be empty'}
          />
          <TextField
            className={styles.surname}
            onChange={handleChange}
            id="surname"
            value={surname}
            label="Surname"
            variant="standard"
            onBlur={validateSurnameField}
            onFocus={suspendSurnameFieldValidation}
            error={!isSurnameValid}
            helperText={isSurnameValid ? ' ' : 'This field should not be empty'}
          />
        </Box>
      </Box>
      <TextField
        label="Phone"
        variant="standard"
        value={phone}
        onChange={handlePhoneNumberChange}
        onBlur={validatePhoneNumber}
        onFocus={suspendPhoneFieldValidation}
        error={!isPhoneNumberValid}
        helperText={isPhoneNumberValid ? ' ' : 'The phone number is invalid'}
      />
    </>
  );
};

PersonalInformation.propTypes = {
  handleChange: PropTypes.func.isRequired,
  setAddress: PropTypes.func.isRequired,
  address: addressType.isRequired,
};

export default PersonalInformation;
