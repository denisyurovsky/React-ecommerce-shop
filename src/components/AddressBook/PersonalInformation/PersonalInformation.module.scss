@import "../../../assets/styles/variables.scss";

.wrapperFullName {
  display: flex;
}

.title {
  flex-basis: 60px;
  flex-shrink: 0;
  margin-right: 10px !important;
}

.name {
  flex-grow: 1;
  margin-right: 10px !important;
}

.surname {
  flex-grow: 1;
}

.wrapper {
  display: flex;
  flex-grow: 1;
}

@media screen and (max-width: $xs) {
  .wrapperFullName {
    flex-wrap: wrap;
  }

  .title {
    margin-right: 0 !important;
    width: 100% !important;
    flex-basis: unset;
  }

  .wrapper {
    margin-top: 10px !important;
  }

  .name {
    flex-grow: 1;
  }

  .surname {
    flex-grow: 1;
  }
}
