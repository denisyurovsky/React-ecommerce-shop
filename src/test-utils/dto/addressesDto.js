export const newAddress = {
  title: 'Mr.',
  name: 'Jon',
  surname: 'Smith',
  country: { id: 'RU', name: 'Russian Federation' },
  city: 'Saint Petersburg',
  street: 'Nevskiy',
  building: '8',
  flat: '33',
  phone: '+7 (981) 1112233',
  zip: '168142',
};

export const addresses = [
  {
    id: 1,
    title: 'Mr.',
    name: 'Alexander',
    surname: 'Sidorov',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pulkovskaya',
    building: '8',
    flat: '4',
    phone: '+7 (981) 3002020',
    zip: '168142',
  },
  {
    id: 2,
    title: 'Mr.',
    name: 'Ivan',
    surname: 'Ivanov',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Leninskiy',
    building: '8',
    flat: '4',
    phone: '+7 (981) 1112233',
    zip: '168142',
  },
  {
    id: 3,
    title: 'Mrs.',
    name: 'Olga',
    surname: 'Petrova',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pobedy',
    building: '8',
    flat: '4',
    phone: '+7 (981) 1233232',
    zip: '168142',
  },
  {
    id: 4,
    ...newAddress,
  },
  {
    id: 5,
    title: 'Mrs.',
    name: 'Elon',
    surname: 'Musk',
    country: { id: 'RU', name: 'Russian Federation' },
    city: 'Saint Petersburg',
    street: 'Nevskiy',
    building: '8',
    flat: '33',
    phone: '+7 (981) 1112233',
    zip: '168142',
    createdAt: '2021-04-10T10:16:28.404Z',
    updatedAt: '2021-12-28T23:24:21.552Z',
  },
];

export const addressesDto = [
  {
    id: 0,
    title: 'Mr.',
    name: 'Alexander',
    surname: 'Sidorov',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pulkovskaya',
    building: '8',
    flat: '4',
    phone: '+7 (981) 3002020',
    zip: '168142',
    createdAt: '2021-04-10T10:16:28.404Z',
    updatedAt: '2021-12-28T23:24:21.552Z',
  },
  {
    id: 2,
    title: 'Mrs.',
    name: 'Alexandra',
    surname: 'Sidorova',
    country: {
      id: 'RU',
      name: 'Russian Federation',
    },
    city: 'Saint Petersburg',
    street: 'Pulkovskaya',
    building: '8',
    flat: '4',
    phone: '+7 (981) 1233232',
    zip: '168142',
    createdAt: '2021-09-07T15:27:43.842Z',
    updatedAt: '2021-10-21T22:59:08.494Z',
  },
];
