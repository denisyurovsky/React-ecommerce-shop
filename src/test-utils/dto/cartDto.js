export const testCart = {
  sellers: {
    1: {
      products: [
        {
          userId: 1,
          productId: 36,
          quantity: 1,
          checked: true,
          price: 724,
          discountPrice: null,
        },
      ],
      checked: true,
    },
    3: {
      products: [
        {
          userId: 3,
          productId: 23,
          quantity: 1,
          checked: true,
          price: 891,
          discountPrice: null,
        },
        {
          userId: 3,
          productId: 28,
          quantity: 1,
          checked: true,
          price: 210,
          discountPrice: 100,
        },
        {
          userId: 3,
          productId: 6,
          quantity: 1,
          checked: true,
          price: 510,
          discountPrice: 400,
        },
      ],
      checked: true,
    },
    4: {
      products: [
        {
          userId: 4,
          productId: 10,
          quantity: 1,
          checked: true,
          price: 757,
          discountPrice: null,
        },
      ],
      checked: true,
    },
  },
  totalPrice: 3092,
  totalDiscountPrice: 2872,
  totalQuantity: 5,
  isLoading: false,
  errorOccurred: false,
};

export const checkoutCart = {
  sellers: {
    1: {
      products: [
        {
          userId: 1,
          productId: 10,
          quantity: 1,
          checked: true,
          price: 917,
          discountPrice: 679,
        },
        {
          userId: 1,
          productId: 24,
          quantity: 1,
          checked: true,
          price: 108,
          discountPrice: 27,
        },
      ],
      checked: true,
    },
    2: {
      products: [
        {
          userId: 2,
          productId: 50,
          quantity: 1,
          checked: true,
          price: 787,
          discountPrice: null,
        },
        {
          userId: 2,
          productId: 33,
          quantity: 2,
          checked: true,
          price: 40,
          discountPrice: null,
        },
      ],
      checked: true,
    },
  },
  totalPrice: 1892,
  totalDiscountPrice: 1573,
  totalQuantity: 5,
  isLoading: false,
  errorOccurred: false,
  errorMessage: '',
};
