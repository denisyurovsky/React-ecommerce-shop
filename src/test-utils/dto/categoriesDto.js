const categoriesDto = [
  'All categories',
  'Grocery',
  'Industrial',
  'Clothing',
  'Computers',
];

export default categoriesDto;
