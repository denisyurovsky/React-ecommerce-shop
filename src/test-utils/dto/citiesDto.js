export const citiesRU = {
  cities: [
    'Krasnolesnyy',
    'Moscow',
    'Indura',
    'Sosnovka',
    'Saint Petersburg',
    'Zykovo',
    'Zyablikovo',
    'Zyukayka',
  ],
};

export const citiesBY = {
  cities: [
    'Kosterevo',
    'Minsk',
    'Sosnovka',
    'Skidel’',
    'Horad Zhodzina',
    'Zhlobin',
    'Zyuzino',
    'Zel’va',
    'Zaslawye',
  ],
};
