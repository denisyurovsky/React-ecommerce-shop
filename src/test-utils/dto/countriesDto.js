export const countries = [
  { id: 'AU', name: 'Australia' },
  { id: 'AT', name: 'Austria' },
  { id: 'BY', name: 'Belarus' },
  { id: 'BE', name: 'Belgium' },
  { id: 'BM', name: 'Bermuda' },
  { id: 'RU', name: 'Russian Federation' },
];

export const objCountries = {
  data: countries,
  isLoading: false,
  errorOccurred: false,
  errorMessage: '',
};
