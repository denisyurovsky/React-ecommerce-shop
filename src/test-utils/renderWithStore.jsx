import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

function render(component, { role = null, store } = {}) {
  let preloadedState;

  if (!store) {
    preloadedState = role
      ? {
          user: {
            user: {
              id: 5,
              role,
            },
          },
        }
      : {};
  }

  const storeWithState = store
    ? store
    : configureStore({
        reducer: {
          cart: cartReducer,
          products: productsReducer,
          categories: categoriesReducer,
          user: userReducer,
          feedback: feedbackReducer,
        },
        preloadedState,
      });

  function Wrapper({ children }) {
    return (
      <Provider store={storeWithState}>
        {children}
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
