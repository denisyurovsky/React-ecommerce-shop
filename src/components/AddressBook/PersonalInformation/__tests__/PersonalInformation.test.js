import React from 'react';

import { addresses } from '../../../../test-utils/dto/addressesDto';
import renderWithStore from '../../../../test-utils/renderWithStore';
import PersonalInformation from '../PersonalInformation';

describe('PersonalInformation component', () => {
  it('should render valid snapshot', () => {
    const { asFragment } = renderWithStore(
      <PersonalInformation
        handleChange={() => {}}
        address={addresses[0]}
        setAddress={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
