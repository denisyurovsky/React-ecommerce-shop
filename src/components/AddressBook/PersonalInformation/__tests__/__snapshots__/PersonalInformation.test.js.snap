// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonalInformation component should render valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapperFullName MuiBox-root css-0"
  >
    <div
      class="MuiFormControl-root title css-1nrlq1o-MuiFormControl-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        id="title"
      >
        Title
      </label>
      <div
        class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
        data-testid="select-title"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="title mui-component-select-title"
          class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          id="mui-component-select-title"
          role="button"
          tabindex="0"
        >
          Mr.
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          name="title"
          tabindex="-1"
          value="Mr."
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="wrapper MuiBox-root css-0"
    >
      <div
        class="MuiFormControl-root MuiTextField-root name css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="name"
          id="name-label"
        >
          Name
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-describedby="name-helper-text"
            aria-invalid="false"
            class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="name"
            type="text"
            value="Alexander"
          />
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
          id="name-helper-text"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </p>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root surname css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="surname"
          id="surname-label"
        >
          Surname
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-describedby="surname-helper-text"
            aria-invalid="false"
            class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="surname"
            type="text"
            value="Sidorov"
          />
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
          id="surname-helper-text"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </p>
      </div>
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="mui-1"
      id="mui-1-label"
    >
      Phone
    </label>
    <div
      class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-describedby="mui-1-helper-text"
        aria-invalid="false"
        class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id="mui-1"
        type="text"
        value="+7 (981) 3002020"
      />
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
      id="mui-1-helper-text"
    >
      <span
        class="notranslate"
      >
        ​
      </span>
    </p>
  </div>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
